source/slang/slang-capability.cpp

```diff
@@ -134,664 +140,144 @@ bool isCapabilityDerivedFrom(CapabilityAtom atom, CapabilityAtom base)
-    return thatIndex == thatCount;
-}
-
-    /// Helper functor for binary search on lists of `CapabilityAtom`
-struct CapabilityAtomComparator
-{
-    int operator()(CapabilityAtom left, CapabilityAtom right)
-    {
-        return int(Int(left) - Int(right));
-    }
-};
-
-bool CapabilityConjunctionSet::implies(CapabilityAtom atom) const
-{
-    // Every non-alias atom that `this` implies should
-    // be presented in the `m_expandedAtoms` list.
-    //
-    // Because the list is sorted, we can find out whether
-    // it contains `atom` with a binary search.
-    //
-    Index result = m_expandedAtoms.binarySearch(atom, CapabilityAtomComparator());
-    return result >= 0;
-}
-
-Int CapabilityConjunctionSet::countIntersectionWith(CapabilityConjunctionSet const& that) const
-{
-    // The goal of this subroutine is to count the number of
-    // elements in the intersection of `this` and `that`,
-    // without explicitly forming that intersection.
-    //
-    // Our approach here will be to iterate over the two
-    // sets in tandem (O(N) in the size of the larger set)
-    // and check for elements that both contain.
-    //
-    // TODO: There should be an asymptotically faster
-    // recursive algorithm here.
-
-    Int intersectionCount = 0;
-
-    Index thisCount = this->m_expandedAtoms.getCount();
-    Index thatCount = that.m_expandedAtoms.getCount();
-    Index thisIndex = 0;
-    Index thatIndex = 0;
-    for(;;)
+    
+    if (m_targetSets.getCount() == 0)
     {
-        if(thisIndex == thisCount) break;
-        if(thatIndex == thatCount) break;
-
-        auto thisAtom = this->m_expandedAtoms[thisIndex];
-        auto thatAtom = that.m_expandedAtoms[thatIndex];
-
-        if( thisAtom == thatAtom )
+        const auto anyTargetInfo = _getInfo(CapabilityName::any_target);
+        auto canonicalRepCount = canonicalRepresentation.getCount();
+        
+        // +2 for textualTarget+TARGET, may also be TARGET only
+        // for this reason we also use 'unsafeSetToCount' to 
+        // allow reuse of the same set in a clean manner.
+        auto maxSizeOfCanonical = canonicalRepCount + 2;
+        List<CapabilityName> listOfTargetAndStageAtoms;
+        listOfTargetAndStageAtoms.setCount(maxSizeOfCanonical); 
+        int iterToAddTo;
+        CapabilityName targetAtom;
+        for (const auto& targetAtomCononicalRep : anyTargetInfo.canonicalRepresentation)
         {
-            // An item both contain.
-
-            intersectionCount++;
-            thisIndex++;
-            thatIndex++;
-            continue;
-        }
-
-        if( thisAtom < thatAtom )
-        {
-            // An item in `this` but not `that`.
-
-            thisIndex++;
-        }
-        else
-        {
-            SLANG_ASSERT(thisAtom > thatAtom);
-
-            // An item in `that` but not `this`.
-
-            thatIndex++;
+            listOfTargetAndStageAtoms.unsafeSetToCount(maxSizeOfCanonical);
```

Comment:
why is this needed? You just called `listOfTargetAndStageAtoms.setCount` before the loop.